src/RobotHub.cpp

```diff
@@ -1,180 +1,449 @@
+//
+// Created by emiel on 22-05-21.
+//
 
-#include <roboteam_utils/Timer.h>
-#include "roboteam_proto/Setting.pb.h"
-
-#include "GRSim.h"
 #include "RobotHub.h"
-#include "SerialDeviceManager.h"
-#include "packing.h"
 
-namespace rtt::robothub {
+#include <cstdio>
+#include <iostream>
+#include <sstream>
+#include <fstream>
+
+#include "LibusbUtilities.h"
+#include "BaseTypes.h"
+#include "RobotCommand.h"
+#include "RobotFeedback.h"
+#include "RobotStateInfo.h"
+#include "Packing.h"
+
+#include "roboteam_proto/State.pb.h"
+
+using namespace std::chrono;
+
+/** @brief Callback funnction triggered when a basestation is connected */
+int hotplug_callback_attach(libusb_context *ctx, libusb_device *device, libusb_hotplug_event event, void *user_data){
+    std::cout << "[hotplug_callback_attach] Event triggered!" << std::endl;
+    auto* robothub = (rtt::robothub::RobotHub*)(user_data);
+    robothub->handleBasestationAttach(device);
+    return 0;
+}
+/** @brief Callback funnction triggered when a basestation is disconnected */
+int hotplug_callback_detach(libusb_context *ctx, libusb_device *device, libusb_hotplug_event event, void *user_data){
+    std::cout << "[hotplug_callback_detach] Event triggered!" << std::endl;
+    auto* robothub = (rtt::robothub::RobotHub*)(user_data);
+    robothub->handleBasestationDetach(device);
+    return 0;
+}
+
+
+namespace rtt {
+namespace robothub {
 
 RobotHub::RobotHub() {
+    std::cout << "[RobotHub] New RobotHub" << std::endl;
     grsimCommander = std::make_shared<GRSimCommander>();
+}
 
-#ifdef __APPLE__
-    device = std::make_shared<SerialDeviceManager>("/dev/cu.usbmodem00000000001A1");
-#else
-    device = std::make_shared<SerialDeviceManager>("/dev/serial/by-id/usb-RTT_BaseStation_00000000001A-if00");
-#endif
-
-    simulator_connection = std::make_shared<SSLSimulator>();
+RobotHub::~RobotHub() {
+    std::cout << "[RobotHub::~RobotHub] Destructor" << std::endl;
 }
 
-/// subscribe to topics
-void RobotHub::subscribeToTopics() {
-    robotCommandSubscriber = new proto::Subscriber<proto::AICommand>(robotCommandChannel, &RobotHub::processAIBatch, this);
+void RobotHub::subscribe(){
+    robotCommandSubscriber = std::make_unique<proto::Subscriber<proto::AICommand>>(
+            proto::ROBOT_COMMANDS_PRIMARY_CHANNEL, &RobotHub::processAIcommand, this
+    );
+
+    worldStateSubscriber = std::make_unique<proto::Subscriber<proto::State>>(
+            proto::WORLD_CHANNEL, &RobotHub::processWorldState, this
+    );
 
-    settingsSubscriber = new proto::Subscriber<proto::Setting>(settingsChannel, &RobotHub::processSettings, this);
+    settingsSubscriber = std::make_unique<proto::Subscriber<proto::Setting>>(
+            proto::SETTINGS_PRIMARY_CHANNEL, &RobotHub::processSettings, this
+    );
 
-    feedbackPublisher = new proto::Publisher<proto::RobotData>(feedbackChannel);
+    feedbackPublisher = std::make_unique<proto::Publisher<proto::RobotFeedback>>(proto::FEEDBACK_PRIMARY_CHANNEL);
```

Comment:
That's doable. Since both `sentCommands` and `receivedFeedback` are repeated fields, how many packages would you like to receive for this? I would still like to replace the handmade RobotFeedback with the REM RobotFeedback